hazelcast/include/hazelcast/util/Future.h

```diff
@@ -40,20 +45,72 @@ namespace hazelcast {
         template<typename T>
         class Future {
         public:
+            class SuccessCallbackRunner : public Runnable {
+            public:
+                SuccessCallbackRunner(T sharedObj,
+                                      const boost::shared_ptr<client::impl::ExecutionCallback<T> > &callback)
+                        : sharedObj(sharedObj), callback(callback) {}
+
+                virtual ~SuccessCallbackRunner() {
+                }
+
+                virtual void run() {
+                    callback->onResponse(&sharedObj);
+                }
+
+                virtual const std::string getName() const {
+                    return "SuccessCallbackRunner";
+                }
+
+            private:
+                T sharedObj;
+                const boost::shared_ptr<client::impl::ExecutionCallback<T> > callback;
+            };
+
+            class ExceptionCallbackRunner : public Runnable {
+            public:
+                ExceptionCallbackRunner(const boost::shared_ptr<client::exception::IException> &exception,
+                                        const boost::shared_ptr<client::impl::ExecutionCallback<T> > &callback)
+                        : exception(exception), callback(callback) {}
+
+                virtual ~ExceptionCallbackRunner() {
+                }
+
+                virtual void run() {
+                    callback->onFailure(exception.get());
+                }
+
+                virtual const std::string getName() const {
+                    return "SuccessCallbackRunner";
+                }
+
+            private:
+                boost::shared_ptr<client::exception::IException> exception;
+                const boost::shared_ptr<client::impl::ExecutionCallback<T> > callback;
+            };
+
             Future()
             : resultReady(false)
             , exceptionReady(false) {
 
             }
 
+            virtual ~Future() {
+            }
+
             void set_value(T& value) {
                 LockGuard guard(mutex);
                 if (exceptionReady || resultReady) {
                     util::ILogger::getLogger().warning(std::string("Future.set_value should not be called twice"));
                 }
                 sharedObject = value;
                 resultReady = true;
+                onComplete();
```

Comment:
What is the use of onComplete? 
On java, this abstraction is necessary because the same invocationFuture is used by both clients and members? 
Here we don't really need that.

Reply:
Java calls callIdSequence.complete(); on completion of ClientInvocationFuture, this is why we need it.